### PR TITLE
Add support for Gorm, Gorestful, Sqlx and Json-iterator

### PR DESCRIPTION
**Pull request by smowton**
This is a cleaned-up version of https://github.com/github/codeql-go/pull/166 --

* Tests added
* Review suggestions from #166 applied
* Library support somewhat generalised (Json-iterator and Go-restful have somewhat broader coverage of the package API)

### CODE REVIEW

File: ql/src/semmle/go/frameworks/Encoding.qll [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  UnmarshalingFunction::Range {
+  JsonIteratorUnmarshalFunction() {
+    this.hasQualifiedName("github.com/json-iterator/go", ["Unmarshal", "UnmarshalFromString"])
+    or
+    exists(Type apiImpl, Type apiType |
+      apiType.hasQualifiedName("github.com/json-iterator/go", "API") and
+      apiImpl.implements(apiType.getUnderlyingType()) and
+      this.(Method).getReceiverType() = apiImpl and
+      this.getName() = ["Unmarshal", "UnmarshalFromString"]
+    )
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/Encoding.qll:18]: Oof, that's a little unfortunate that we need this. Does something like this work instead?

```suggestion
    exists(Method m |
      m.hasQualifiedName("github.com/json-iterator/go", "API", ["Unmarshal", "UnmarshalFromString"]) and
      this.(Method).implements(m) 
    )
```

File: ql/src/semmle/go/frameworks/HTTP.qll [hunk 2]
```diff
@@ -231,3 +231,44 @@ private module StdlibHttp {
     }
   }
 }
+
+/**
+ * Models of the go-restful library (https://github.com/emicklei/go-restful)
+ */
+private module GoRestfulHttp {
+  /**
+   * Methods defined on go-restful's Request object that may return user-controlled data.
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/HTTP.qll:240]:
```suggestion
   * A model for methods defined on go-restful's `Request` object that may return user-controlled data.
```

File: ql/src/semmle/go/frameworks/SQL.qll [hunk 3]
```diff
@@ -160,4 +160,29 @@ module SQL {
       }
     }
   }
+
+  /** Sinks of github.com/jinzhu/gorm */
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/SQL.qll:164]:
```suggestion
  /** A model for sinks of github.com/jinzhu/gorm. */
```

File: ql/src/semmle/go/frameworks/SQL.qll [hunk 4]
```diff
@@ -160,4 +160,29 @@ module SQL {
...
+    GormSink() {
+      exists(Method meth, string name |
+        meth.hasQualifiedName("github.com/jinzhu/gorm", "DB", name) and
+        this = meth.getACall().getArgument(0) and
+        name = ["Where", "Raw", "Order", "Not", "Or", "Select", "Table", "Group", "Having", "Joins"]
+      )
+    }
+  }
+
+  /** Sinks of github.com/jmoiron/sqlx */
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/SQL.qll:175]:
```suggestion
  /** A model for sinks of github.com/jmoiron/sqlx. */
```

File: ql/src/semmle/go/frameworks/SQL.qll [hunk 5]
```diff
@@ -160,4 +160,29 @@ module SQL {
       }
     }
   }
+
+  /** Sinks of github.com/jinzhu/gorm */
+  private class GormSink extends DataFlow::Node, SQL::QueryString::Range {
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/SQL.qll:165]:
```suggestion
  private class GormSink extends SQL::QueryString::Range {
```

File: ql/src/semmle/go/frameworks/SQL.qll [hunk 6]
```diff
@@ -160,4 +160,29 @@ module SQL {
...
+      exists(Method meth, string name |
+        meth.hasQualifiedName("github.com/jinzhu/gorm", "DB", name) and
+        this = meth.getACall().getArgument(0) and
+        name = ["Where", "Raw", "Order", "Not", "Or", "Select", "Table", "Group", "Having", "Joins"]
+      )
+    }
+  }
+
+  /** Sinks of github.com/jmoiron/sqlx */
+  private class SqlxSink extends DataFlow::Node, SQL::QueryString::Range {
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/SQL.qll:176]:
```suggestion
  private class SqlxSink extends SQL::QueryString::Range {
```

File: ql/src/semmle/go/frameworks/HTTP.qll [hunk 7]
```diff
@@ -231,3 +231,44 @@ private module StdlibHttp {
...
+            "Request",
+            ["QueryParameters", "QueryParameter", "BodyParameter", "HeaderParameter",
+                "PathParameter", "PathParameters"])
+    }
+  }
+
+  /**
+   * Models go-restful's Request object as a source of user-controlled data.
+   */
+  private class GoRestfulSource extends DataFlow::Node, UntrustedFlowSource::Range {
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/HTTP.qll:255]:
```suggestion
  private class GoRestfulSource extends UntrustedFlowSource::Range {
```

**Comment by smowton** [comment]
@max-schaefer changes applied, change-note added.

**Comment by max-schaefer** [comment]
LGTM; have we done a dist-compare for this already?

**Comment by smowton** [comment]
Running presently, results in the morning

**Pull request merged**